apps/web/src/components/Auth/TwitterAuth.tsx

```diff
@@ -0,0 +1,103 @@
+import { useRouter } from 'next/router';
+import { useCallback, useEffect } from 'react';
+import { useFragment } from 'react-relay';
+import { graphql } from 'relay-runtime';
+
+import { useToastActions } from '~/contexts/toast/ToastContext';
+import { TwitterAuthMutation } from '~/generated/TwitterAuthMutation.graphql';
+import { TwitterAuthQueryFragment$key } from '~/generated/TwitterAuthQueryFragment.graphql';
+import { usePromisifiedMutation } from '~/hooks/usePromisifiedMutation';
+
+import FullPageLoader from '../core/Loader/FullPageLoader';
+
+type Props = {
+  queryRef: TwitterAuthQueryFragment$key;
+};
+
+export default function TwitterAuth({ queryRef }: Props) {
+  const query = useFragment(
+    graphql`
+      fragment TwitterAuthQueryFragment on Query {
+        viewer {
+          ... on Viewer {
+            user {
+              username
+            }
+          }
+        }
+      }
+    `,
+    queryRef
+  );
+
+  const username = query.viewer?.user?.username;
+
+  const [verifyTwitter] = usePromisifiedMutation<TwitterAuthMutation>(graphql`
+    mutation TwitterAuthMutation($input: SocialAuthMechanism!) {
+      connectSocialAccount(input: $input) {
+        __typename
+      }
+    }
+  `);
+
+  const router = useRouter();
+  const { pushToast } = useToastActions();
+
+  const { code, error } = router.query;
+
+  if (!username) {
+    throw new Error('Try to authorize Twitter without username');
+  }
+
+  const handleVerifyTwitter = useCallback(async () => {
+    if (!code) return;
+
+    const payload = {
+      twitter: {
+        code: code as string,
+      },
+    };
+
+    try {
+      await verifyTwitter({
+        variables: {
+          input: payload,
+        },
+      });
+
+      router.push({
+        pathname: '/[username]',
+        query: {
+          username,
+        },
+      });
+    } catch (error) {
+      if (error instanceof Error) {
+        pushToast({
+          message: 'Could not authorize Gallery on Twitter',
+        });
+      }
+    }
+  }, [code, pushToast, router, username, verifyTwitter]);
+
+  useEffect(() => {
+    handleVerifyTwitter();
+  }, [handleVerifyTwitter]);
+
+  useEffect(() => {
+    if (error) {
+      router.push({
+        pathname: '/[username]',
+        query: {
+          username,
+        },
+      });
+
+      pushToast({
+        message: 'Could not authorize Gallery on Twitter',
+      });
+    }
+  }, [error, pushToast, router, username]);
```

Comment:
it seems unusual that we always call `handleVerifyTwitter` AND do the redirect if there's an error. i think it should be an if-else

Reply:
handled here: https://github.com/gallery-so/gallery/pull/1343/commits/190ce8781b7fe0167bdcdbd39188ecfd782be849